administrator/components/com_installer/Model/Database.php

```diff
@@ -41,61 +68,126 @@ class Database extends Installer
+	public function fix($extensionIdArray = null)
 	{
-		if (!$changeSet = $this->getItems())
+		if (!$changeSetList = $this->getItems($extensionIdArray))
 		{
 			return false;
 		}
 
-		$changeSet->fix();
-		$this->fixSchemaVersion($changeSet);
-		$this->fixUpdateVersion();
-		$installer = new \JoomlaInstallerScript;
-		$installer->deleteUnexistingFiles();
-		$this->fixDefaultTextFilters();
-
-		/*
-		 * Finally, if the schema updates succeeded, make sure the database is
-		 * converted to utf8mb4 or, if not suported by the server, compatible to it.
-		 */
-		$statusArray = $changeSet->getStatus();
+		foreach ($changeSetList as $i => $changeSet)
+		{
+			$changeSet['changeset']->fix();
+			$this->fixSchemaVersion($changeSet['changeset'], $changeSet['extension']->extension_id);
+		}
 
-		if (count($statusArray['error']) == 0)
+		// If null it means it is Joomla! core database
+		if ($extensionIdArray == null)
 		{
-			$installer->convertTablesToUtf8mb4(false);
+			$this->fixUpdateVersion();
+			$installer = new \JoomlaInstallerScript;
+			$installer->deleteUnexistingFiles();
+			$this->fixDefaultTextFilters();
+
+			/*
+			 * Finally, if the schema updates succeeded, make sure the database is
+			 * converted to utf8mb4 or, if not suported by the server, compatible to it.
+			 */
+			$statusArray = $changeSet['changeset']->getStatus();
+
+			if (count($statusArray['error']) == 0)
+			{
+				$installer->convertTablesToUtf8mb4(false);
+			}
 		}
+
 	}
 
 	/**
 	 * Gets the changeset object.
 	 *
+	 * @param   array  $extensionIdArray  list of the selected extensions to fix
+	 *
 	 * @return  \Joomla\CMS\Schema\ChangeSet
 	 */
-	public function getItems()
+	public function getItems($extensionIdArray = null)
 	{
-		$folder = JPATH_ADMINISTRATOR . '/components/com_admin/sql/updates/';
+		$session = \JFactory::getSession();
+
+		$changeSetList = $session->get('changeSetList');
+
+		if ($changeSetList != null)
+		{
+			return json_decode($changeSetList);
+		}
+
+		$db    = $this->getDbo();
+		$query = $db->getQuery(true)
+			->select('*')
+			->from(
+				$db->quoteName(
+					'#__schemas',
+					's'
+				)
+			)->join(
+				'INNER',
+				$db->quoteName(
+					'#__extensions', 'e'
+				) . ' ON (' . $db->quoteName(
+					's.extension_id'
+				) . ' = ' . $db->quoteName(
+					'e.extension_id'
+				) . ')'
+			);
+
+		if ($extensionIdArray != null)
+		{
+			$whereQuery = array();
+
+			foreach ($extensionIdArray as $extension)
+			{
+				array_push($whereQuery, 'e.extension_id = ' . $extension);
+			}
+
+			$query->where($whereQuery);
+		}
+
+		$result = $this->_getList($query);
+
+		$changeSetList = array();
 
 		try
 		{
-			$changeSet = ChangeSet::getInstance($this->getDbo(), $folder);
+			foreach ($result as $index => $result)
+			{
+				if (strcmp($result->element, 'joomla') == 0)
+				{
+					$result->element = 'com_admin';
+					$index = 'core';
+				}
+
+				$folderTmp = JPATH_ADMINISTRATOR . '/components/' . $result->element . '/sql/updates/';
```

Comment:
this is too much hardcoded

com_patchtester for example  does'nt follow the same path